zarr_sparse/buffer.py

```diff
@@ -0,0 +1,291 @@
+from __future__ import annotations
+
+import math
+from typing import TYPE_CHECKING, Any, Literal
+
+import numpy as np
+import numpy.typing as npt
+import sparse
+from zarr.core.buffer.core import Buffer, BufferPrototype, NDBuffer
+from zarr.registry import register_ndbuffer
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+    from typing import Self
+
+
+def slice_size(slice_, size):
+    return len(range(*slice_.indices(size)))
+
+
+def normalize_slice(slice_, size):
+    return slice(*slice_.indices(size))
+
+
+def combine_nd(parts):
+    raise NotImplementedError
+
+
+def expand_chunks(chunks, shape):
+    def _expand(chunkspec, size):
+        if chunkspec == -1:
+            return (size,)
+        elif isinstance(chunkspec, int):
+            n_full_chunks, remainder = divmod(size, chunkspec)
+            chunks = (chunkspec,) * n_full_chunks
+
+            if remainder > 0:
+                chunks += (remainder,)
+
+            return chunks
+
+        chunkspec = tuple(chunkspec)
+        if sum(chunkspec) != size:
+            raise ValueError(f"chunks don't add up to the full size: {chunkspec}")
+
+        return chunkspec
+
+    return tuple(_expand(chunkspec, size) for chunkspec, size in zip(chunks, shape))
+
+
+def slice_to_chunk_index(slice_, offsets):
+    if slice_ == slice(None):
+        return 0
+
+    chunk_index = np.flatnonzero(offsets == slice_.start)
+
+    if chunk_index.size == 0:
+        raise ValueError(f"Selected chunk out of bounds: {slice_}")
+
+    return chunk_index[0]
+
+
+def sparse_equal(a, b, equal_nan: bool) -> bool:
+    equal_nan = equal_nan if a.dtype.kind not in ("U", "S", "T", "O", "V") else False
+
+    if b.ndim == 0:
+        if not np.array_equal(
+            a.fill_value, getattr(b, "fill_value", b), equal_nan=equal_nan
+        ):
+            return False
+
+        if a.nnz == 0:
+            return True
+
+        return np.array_equal(
+            a.data,
+            np.broadcast_to(b.data, a.data.shape),
+            equal_nan=equal_nan,
+        )
+
+    # use array_equal to obtain equal_nan=True functionality
+    # Since fill-value is a scalar, isn't there a faster path than allocating a new array for fill value
+    # every single time we have to write data?
+    _data, other = sparse.broadcast_arrays(a, b)
+
+    return sparse.equal(a, other, equal_nan=equal_nan)
+
+
+class ChunkGrid:
```

Comment:
This is a quite interesting abstraction. I was toying around with something similar in https://github.com/zarr-developers/zarr-python/issues/3175#issuecomment-3127987591 but I think this could be a simpler way of handling irregular grids (in this case, output type, in the linked case, output shape)